python/src/space/core/runners.py

```diff
@@ -123,6 +123,35 @@ def decorated(self, *args, **kwargs):
 
     return decorated
 
+class JobExecutionUtils:
+  """Untils class for utils methods for job execution."""
+  @staticmethod
+  def wrap_error(fn: Callable[...,JobResult]) -> Callable[..., JobResult]:
+    """A decorator that wrap error."""
+
+    @wraps(fn)
+    def decorated(self, *args, **kwargs):
+      try:
+        r = fn(self, *args, **kwargs)
+        logging.info(f"Job result:\n{r}")
+        return r
+      except (errors.SpaceRuntimeError, errors.UserInputError) as e:
+        r = JobResult(JobResult.State.FAILED, None, repr(e))
+        logging.warning(f"Job result:\n{r}")
+        return r
+    return decorated
+
+class BaseVersionUpdateRunner(ABC):
+  """Abstract base version management runner class."""
+
+  @abstractmethod
+  def add_tag(self, tag:str, snapshot_id: Optional[int] = None) -> JobResult:
```

Comment:
Can we directly let dataset support tags, instead of runner? I feel it is not necessary to use runner here. Runner is for expensive operations (when need to select between local or distributed runner).

Can we do ds.add_tag and ds.remove_tag directly?

Reply:
Done